mlem/contrib/lightgbm.py

```diff
@@ -86,33 +106,51 @@ def write(
             raise ValueError(
                 f"expected data to be of LightGBMDataType, got {type(data)} instead"
             )
+        res = {}
         lightgbm_construct = data.data.construct()
         raw_data = lightgbm_construct.get_data()
-        underlying_labels = lightgbm_construct.get_label().tolist()
-        inner_reader, art = data.inner.get_writer().write(
-            data.inner.copy().bind(raw_data), storage, path
+        raw_label = lightgbm_construct.get_label().tolist()
+
+        inner_reader, inner_art = data.inner.get_writer().write(
+            data.inner.copy().bind(raw_data),
+            storage,
+            posixpath.join(path, "inner"),
+        )
+        labels_reader, labels_art = data.labels.get_writer().write(
```

Comment:
Can we do old logic if no labels present?
Without labels: `save(data, "dataset")` produces
```
dataset.mlem
dataset
```
With labels:
```
dataset.mlem
dataset/data
dataset/labels
```

Reply:
I think I made it work, unfortunately, lgb tests do not work for me locally, will check on CI

Reply:
Hmm, change of implementation, using `.construct()` now
this ensures that labels are never None and always exist

Reply:
We will always have labels now, resolving

Reply:
We should not modify the object passed to `process`. And we should always have exactly the same object when we `load` something we `save`d before. Is this the case if you are using `construct`?

Reply:
When you instantiate a `lgb.Dataset`, you cannot call `get_data` and `get_label` on it since it is lazy initialized i.e. it hasn't been constructed. We were using `get_data` before this PR also i.e. we were using `.construct` in the `writer`.

Further, when you pass `labels` as None, the `.construct()` operation sets them to some values i.e. it is never None after it has been constructed.

The good thing about `.construct` is that when it is called on an already `constructed` object -- it makes no difference i.e. it can be called infinitely more times.

And methods like `lgb.train` also construct it under the hood.

So do you want to work with the raw object OR the constructed object.
If the RAW object, then I will have to change the PR so that `.construct` is never called anywhere aka the existing implementation of this PR was also kinda wrong then.

If the Constructed object, then this PR is kinda ready as it is.

Reply:
I have removed the usage of `construct` now

Reply:
Ok, now it looks a lot better - basically labels are just a second dataset.
Now again, if you save an `lgb.Dataset` without labels, can we  do it with a single file like I proposed earlier?
Basically if labels are not `None` this should be the same logic. If they are, you'll have `labels` as `None` (in reader/writer as well) and just one artifact at `<dataset>` instead of `<dataset>/data`

Reply:
I have accommodated that `labels` should not be saved when it is `None`.
But then, the key with it is saved as, is still `<dataset>/data`
Do you really wanna have `<dataset>` instead of `<dataset>/data`? I just ask because it will lead to many if/else branches I guess.

Reply:
nvm, I made it `<dataset>` instead of `<dataset>/data` 😄